# Computer Vision Nanodegree

## Project: Image Captioning

---

In this notebook, you will train your CNN-RNN model.  

You are welcome and encouraged to try out many different architectures and hyperparameters when searching for a good model.

This does have the potential to make the project quite messy!  Before submitting your project, make sure that you clean up:
- the code you write in this notebook.  The notebook should describe how to train a single CNN-RNN architecture, corresponding to your final choice of hyperparameters.  You should structure the notebook so that the reviewer can replicate your results by running the code in this notebook.  
- the output of the code cell in **Step 2**.  The output should show the output obtained when training the model from scratch.

This notebook **will be graded**.  

Feel free to use the links below to navigate the notebook:
- [Step 1](#step1): Training Setup
- [Step 2](#step2): Train your Model
- [Step 3](#step3): (Optional) Validate your Model

<a id='step1'></a>
## Step 1: Training Setup

In this step of the notebook, you will customize the training of your CNN-RNN model by specifying hyperparameters and setting other options that are important to the training procedure.  The values you set now will be used when training your model in **Step 2** below.

You should only amend blocks of code that are preceded by a `TODO` statement.  **Any code blocks that are not preceded by a `TODO` statement should not be modified**.

### Task #1

Begin by setting the following variables:
- `batch_size` - the batch size of each training batch.  It is the number of image-caption pairs used to amend the model weights in each training step. 
- `vocab_threshold` - the minimum word count threshold.  Note that a larger threshold will result in a smaller vocabulary, whereas a smaller threshold will include rarer words and result in a larger vocabulary.  
- `vocab_from_file` - a Boolean that decides whether to load the vocabulary from file. 
- `embed_size` - the dimensionality of the image and word embeddings.  
- `hidden_size` - the number of features in the hidden state of the RNN decoder.  
- `num_epochs` - the number of epochs to train the model.  We recommend that you set `num_epochs=3`, but feel free to increase or decrease this number as you wish.  [This paper](https://arxiv.org/pdf/1502.03044.pdf) trained a captioning model on a single state-of-the-art GPU for 3 days, but you'll soon see that you can get reasonable results in a matter of a few hours!  (_But of course, if you want your model to compete with current research, you will have to train for much longer._)
- `save_every` - determines how often to save the model weights.  We recommend that you set `save_every=1`, to save the model weights after each epoch.  This way, after the `i`th epoch, the encoder and decoder weights will be saved in the `models/` folder as `encoder-i.pkl` and `decoder-i.pkl`, respectively.
- `print_every` - determines how often to print the batch loss to the Jupyter notebook while training.  Note that you **will not** observe a monotonic decrease in the loss function while training - this is perfectly fine and completely expected!  You are encouraged to keep this at its default value of `100` to avoid clogging the notebook, but feel free to change it.
- `log_file` - the name of the text file containing - for every step - how the loss and perplexity evolved during training.

If you're not sure where to begin to set some of the values above, you can peruse [this paper](https://arxiv.org/pdf/1502.03044.pdf) and [this paper](https://arxiv.org/pdf/1411.4555.pdf) for useful guidance!  **To avoid spending too long on this notebook**, you are encouraged to consult these suggested research papers to obtain a strong initial guess for which hyperparameters are likely to work best.  Then, train a single model, and proceed to the next notebook (**3_Inference.ipynb**).  If you are unhappy with your performance, you can return to this notebook to tweak the hyperparameters (and/or the architecture in **model.py**) and re-train your model.

### Question 1

**Question:** Describe your CNN-RNN architecture in detail.  With this architecture in mind, how did you select the values of the variables in Task 1?  If you consulted a research paper detailing a successful implementation of an image captioning model, please provide the reference.

**Answer:**  Based on the paper titled "Show and Tell: A Neural Image Caption Generator" i choose to have an Encoder - Decoder Architecture, we start by Encoder which is a pre-trained CNN with a classification task, then we remove the last layer and we add a fully connected layer, wish is the features encoder, then we concatinate it to an Decoder, which a non-trained LSTM which result each time a one-hot vector with more than 8000 value, the value are mapped to a dictionary of words predefined already. than we apply a arg-sigmoid to get the predicted world each time.
For hyper-parameters; i started by basic values as 256 for embeded size, 512 for hidden size and 32 as batch size. as all of them fit well in the memory. and based on the experiental my model performed well and i found that i can get good result ouf of them. 


### (Optional) Task #2

Note that we have provided a recommended image transform `transform_train` for pre-processing the training images, but you are welcome (and encouraged!) to modify it as you wish.  When modifying this transform, keep in mind that:
- the images in the dataset have varying heights and widths, and 
- if using a pre-trained model, you must perform the corresponding appropriate normalization.

### Question 2

**Question:** How did you select the transform in `transform_train`?  If you left the transform at its provided value, why do you think that it is a good choice for your CNN architecture?

**Answer:** I kept the parameters as they are:

The task is to describe a photo so i don't think we can downsample the image more than 256x256, if we do more we will lose details
i keept the image treatments as : RandomCrop, RandomHorizontalFlip, Normalize as they will be used as data augmentation and avoiding overfiting.
as batch size i putted 32, it is within the recommended range.

### Task #3

Next, you will specify a Python list containing the learnable parameters of the model.  For instance, if you decide to make all weights in the decoder trainable, but only want to train the weights in the embedding layer of the encoder, then you should set `params` to something like:
```
params = list(decoder.parameters()) + list(encoder.embed.parameters()) 
```

### Question 3

**Question:** How did you select the trainable parameters of your architecture?  Why do you think this is a good choice?

**Answer:** For the CNN , it is taken from a pre-trained model so we don't have to restart from scratch. i only changed the last layer by fully connected layer to adapt it to the LSTM, the FC will be the feature detector of our CNN. 
Fie the RNN, i trained it from scratch, so all the weights in the decoder was trained based in our new data.

### Task #4

Finally, you will select an [optimizer](http://pytorch.org/docs/master/optim.html#torch.optim.Optimizer).

### Question 4

**Question:** How did you select the optimizer used to train your model?

**Answer:** i used Adam optimiser as it fix well the overfiting problem

In [12]:
import torch
import torch.nn as nn
from torchvision import transforms
import sys
sys.path.append('/opt/cocoapi/PythonAPI')
from pycocotools.coco import COCO
from data_loader import get_loader
from model import EncoderCNN, DecoderRNN
import math


## TODO #1: Select appropriate values for the Python variables below.
batch_size = 64          # batch size
vocab_threshold = 5        # minimum word count threshold
vocab_from_file = True    # if True, load existing vocab file
embed_size = 512           # dimensionality of image and word embeddings
hidden_size = 512          # number of features in hidden state of the RNN decoder
num_epochs = 3             # number of training epochs
save_every = 1             # determines frequency of saving model weights
print_every = 50          # determines window for printing average loss
log_file = 'training_log.txt'       # name of file with saved training loss and perplexity

# (Optional) TODO #2: Amend the image transform below.
transform_train = transforms.Compose([ 
    transforms.Resize(256),                          # smaller edge of image resized to 256
    transforms.RandomCrop(224),                      # get 224x224 crop from random location
    transforms.RandomHorizontalFlip(),               # horizontally flip image with probability=0.5
    transforms.ToTensor(),                           # convert the PIL Image to a tensor
    transforms.Normalize((0.485, 0.456, 0.406),      # normalize image for pre-trained model
                         (0.229, 0.224, 0.225))])

# Build data loader.
data_loader = get_loader(transform=transform_train,
                         mode='train',
                         batch_size=batch_size,
                         vocab_threshold=vocab_threshold,
                         vocab_from_file=vocab_from_file)

# The size of the vocabulary.
vocab_size = len(data_loader.dataset.vocab)

# Initialize the encoder and decoder. 
encoder = EncoderCNN(embed_size)
decoder = DecoderRNN(embed_size, hidden_size, vocab_size)

# Move models to GPU if CUDA is available. 
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
encoder.to(device)
decoder.to(device)

# Define the loss function. 
criterion = nn.CrossEntropyLoss().cuda() if torch.cuda.is_available() else nn.CrossEntropyLoss()

# TODO #3: Specify the learnable parameters of the model.
params = list(decoder.parameters()) + list(encoder.embed.parameters()) 

# TODO #4: Define the optimizer.
# criterion = nn.MSELoss()

# optimizer = torch.optim.Adam(params, lr=0.001, betas=(0.9, 0.999), eps=1e-08)
optimizer = torch.optim.Adam(params)

# Set the total number of training steps per epoch.
total_step = math.ceil(len(data_loader.dataset.caption_lengths) / data_loader.batch_sampler.batch_size)

Vocabulary successfully loaded from vocab.pkl file!
loading annotations into memory...
Done (t=0.60s)
creating index...


  0%|          | 796/414113 [00:00<00:51, 7953.14it/s]

index created!
Obtaining caption lengths...


100%|██████████| 414113/414113 [00:46<00:00, 8993.10it/s]


<a id='step2'></a>
## Step 2: Train your Model

Once you have executed the code cell in **Step 1**, the training procedure below should run without issue.  

It is completely fine to leave the code cell below as-is without modifications to train your model.  However, if you would like to modify the code used to train the model below, you must ensure that your changes are easily parsed by your reviewer.  In other words, make sure to provide appropriate comments to describe how your code works!  

You may find it useful to load saved weights to resume training.  In that case, note the names of the files containing the encoder and decoder weights that you'd like to load (`encoder_file` and `decoder_file`).  Then you can load the weights by using the lines below:

```python
# Load pre-trained weights before resuming training.
encoder.load_state_dict(torch.load(os.path.join('./models', encoder_file)))
decoder.load_state_dict(torch.load(os.path.join('./models', decoder_file)))
```

While trying out parameters, make sure to take extensive notes and record the settings that you used in your various training runs.  In particular, you don't want to encounter a situation where you've trained a model for several hours but can't remember what settings you used :).

### A Note on Tuning Hyperparameters

To figure out how well your model is doing, you can look at how the training loss and perplexity evolve during training - and for the purposes of this project, you are encouraged to amend the hyperparameters based on this information.  

However, this will not tell you if your model is overfitting to the training data, and, unfortunately, overfitting is a problem that is commonly encountered when training image captioning models.  

For this project, you need not worry about overfitting. **This project does not have strict requirements regarding the performance of your model**, and you just need to demonstrate that your model has learned **_something_** when you generate captions on the test data.  For now, we strongly encourage you to train your model for the suggested 3 epochs without worrying about performance; then, you should immediately transition to the next notebook in the sequence (**3_Inference.ipynb**) to see how your model performs on the test data.  If your model needs to be changed, you can come back to this notebook, amend hyperparameters (if necessary), and re-train the model.

That said, if you would like to go above and beyond in this project, you can read about some approaches to minimizing overfitting in section 4.3.1 of [this paper](http://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=7505636).  In the next (optional) step of this notebook, we provide some guidance for assessing the performance on the validation dataset.

In [13]:
import torch.utils.data as data
import numpy as np
import os
import requests
import time

# Open the training log file.
f = open(log_file, 'w')

old_time = time.time()
# response = requests.request("GET", 
#                             "http://metadata.google.internal/computeMetadata/v1/instance/attributes/keep_alive_token", 
#                             headers={"Metadata-Flavor":"Google"})

for epoch in range(1, num_epochs+1):
    
    for i_step in range(1, total_step+1):
        
        if time.time() - old_time > 60:
            old_time = time.time()
#             requests.request("POST", 
#                              "https://nebula.udacity.com/api/v1/remote/keep-alive", 
#                              headers={'Authorization': "STAR " + response.text})
        
        # Randomly sample a caption length, and sample indices with that length.
        indices = data_loader.dataset.get_train_indices()
        # Create and assign a batch sampler to retrieve a batch with the sampled indices.
        new_sampler = data.sampler.SubsetRandomSampler(indices=indices)
        data_loader.batch_sampler.sampler = new_sampler
        
        # Obtain the batch.
        images, captions = next(iter(data_loader))

        # Move batch of images and captions to GPU if CUDA is available.
        images = images.to(device)
        captions = captions.to(device)
        
        # Zero the gradients.
        decoder.zero_grad()
        encoder.zero_grad()
        
        # Pass the inputs through the CNN-RNN model.
        features = encoder(images)
        outputs = decoder(features, captions)
        
        # Calculate the batch loss.
        loss = criterion(outputs.view(-1, vocab_size), captions.view(-1))
        
        # Backward pass.
        loss.backward()
        
        # Update the parameters in the optimizer.
        optimizer.step()
            
        # Get training statistics.
        stats = 'Epoch [%d/%d], Step [%d/%d], Loss: %.4f, Perplexity: %5.4f' % (epoch, num_epochs, i_step, total_step, loss.item(), np.exp(loss.item()))
        
        # Print training statistics (on same line).
        print('\r' + stats, end="")
        sys.stdout.flush()
        
        # Print training statistics to file.
        f.write(stats + '\n')
        f.flush()
        
        # Print training statistics (on different line).
        if i_step % print_every == 0:
            print('\r' + stats)
            
    # Save the weights.
    if epoch % save_every == 0:
        torch.save(decoder.state_dict(), os.path.join('./models', 'decoder-%d.pkl' % epoch))
        torch.save(encoder.state_dict(), os.path.join('./models', 'encoder-%d.pkl' % epoch))

# Close the training log file.
f.close()

Epoch [1/3], Step [50/6471], Loss: 4.3842, Perplexity: 80.176392
Epoch [1/3], Step [100/6471], Loss: 4.1032, Perplexity: 60.5340
Epoch [1/3], Step [150/6471], Loss: 3.7305, Perplexity: 41.70008
Epoch [1/3], Step [200/6471], Loss: 3.2497, Perplexity: 25.78201
Epoch [1/3], Step [250/6471], Loss: 3.2633, Perplexity: 26.1351
Epoch [1/3], Step [300/6471], Loss: 3.1209, Perplexity: 22.6657
Epoch [1/3], Step [350/6471], Loss: 3.2773, Perplexity: 26.5033
Epoch [1/3], Step [400/6471], Loss: 2.8856, Perplexity: 17.9136
Epoch [1/3], Step [450/6471], Loss: 3.0122, Perplexity: 20.3323
Epoch [1/3], Step [500/6471], Loss: 3.1713, Perplexity: 23.8389
Epoch [1/3], Step [550/6471], Loss: 2.9168, Perplexity: 18.4819
Epoch [1/3], Step [600/6471], Loss: 3.2073, Perplexity: 24.7120
Epoch [1/3], Step [650/6471], Loss: 2.6989, Perplexity: 14.8640
Epoch [1/3], Step [700/6471], Loss: 3.2675, Perplexity: 26.2447
Epoch [1/3], Step [750/6471], Loss: 3.0399, Perplexity: 20.9041
Epoch [1/3], Step [800/6471], Loss: 2

Epoch [1/3], Step [6350/6471], Loss: 2.1119, Perplexity: 8.26420
Epoch [1/3], Step [6400/6471], Loss: 2.1100, Perplexity: 8.24801
Epoch [1/3], Step [6450/6471], Loss: 1.8935, Perplexity: 6.64237
Epoch [2/3], Step [50/6471], Loss: 1.8827, Perplexity: 6.5715843
Epoch [2/3], Step [100/6471], Loss: 2.2615, Perplexity: 9.5970
Epoch [2/3], Step [150/6471], Loss: 2.2782, Perplexity: 9.75944
Epoch [2/3], Step [200/6471], Loss: 2.4260, Perplexity: 11.3137
Epoch [2/3], Step [250/6471], Loss: 2.0697, Perplexity: 7.92274
Epoch [2/3], Step [300/6471], Loss: 2.0357, Perplexity: 7.65769
Epoch [2/3], Step [350/6471], Loss: 2.3026, Perplexity: 10.0000
Epoch [2/3], Step [400/6471], Loss: 1.9954, Perplexity: 7.35487
Epoch [2/3], Step [450/6471], Loss: 2.0060, Perplexity: 7.43333
Epoch [2/3], Step [500/6471], Loss: 2.5201, Perplexity: 12.4301
Epoch [2/3], Step [550/6471], Loss: 2.5508, Perplexity: 12.8176
Epoch [2/3], Step [600/6471], Loss: 1.9698, Perplexity: 7.16969
Epoch [2/3], Step [650/6471], Loss: 1

Epoch [2/3], Step [6200/6471], Loss: 1.6827, Perplexity: 5.37994
Epoch [2/3], Step [6250/6471], Loss: 1.9126, Perplexity: 6.77081
Epoch [2/3], Step [6300/6471], Loss: 1.6225, Perplexity: 5.06552
Epoch [2/3], Step [6350/6471], Loss: 1.8170, Perplexity: 6.15332
Epoch [2/3], Step [6400/6471], Loss: 1.9440, Perplexity: 6.98651
Epoch [2/3], Step [6450/6471], Loss: 1.9625, Perplexity: 7.11702
Epoch [3/3], Step [50/6471], Loss: 2.2444, Perplexity: 9.434515
Epoch [3/3], Step [100/6471], Loss: 1.9115, Perplexity: 6.7630
Epoch [3/3], Step [150/6471], Loss: 2.2255, Perplexity: 9.25793
Epoch [3/3], Step [200/6471], Loss: 2.1683, Perplexity: 8.74320
Epoch [3/3], Step [250/6471], Loss: 2.2393, Perplexity: 9.38693
Epoch [3/3], Step [300/6471], Loss: 2.0512, Perplexity: 7.77758
Epoch [3/3], Step [350/6471], Loss: 2.9108, Perplexity: 18.3721
Epoch [3/3], Step [400/6471], Loss: 1.9584, Perplexity: 7.08816
Epoch [3/3], Step [450/6471], Loss: 1.9230, Perplexity: 6.84136
Epoch [3/3], Step [500/6471], Loss:

Epoch [3/3], Step [6050/6471], Loss: 1.9303, Perplexity: 6.89133
Epoch [3/3], Step [6100/6471], Loss: 1.9546, Perplexity: 7.06130
Epoch [3/3], Step [6150/6471], Loss: 1.6464, Perplexity: 5.18802
Epoch [3/3], Step [6200/6471], Loss: 1.7775, Perplexity: 5.91502
Epoch [3/3], Step [6250/6471], Loss: 1.9214, Perplexity: 6.83050
Epoch [3/3], Step [6300/6471], Loss: 1.8363, Perplexity: 6.2736
Epoch [3/3], Step [6350/6471], Loss: 1.8851, Perplexity: 6.58709
Epoch [3/3], Step [6400/6471], Loss: 1.9448, Perplexity: 6.9925
Epoch [3/3], Step [6450/6471], Loss: 2.0632, Perplexity: 7.87155
Epoch [3/3], Step [6471/6471], Loss: 1.9122, Perplexity: 6.76839

<a id='step3'></a>
## Step 3: (Optional) Validate your Model

To assess potential overfitting, one approach is to assess performance on a validation set.  If you decide to do this **optional** task, you are required to first complete all of the steps in the next notebook in the sequence (**3_Inference.ipynb**); as part of that notebook, you will write and test code (specifically, the `sample` method in the `DecoderRNN` class) that uses your RNN decoder to generate captions.  That code will prove incredibly useful here. 

If you decide to validate your model, please do not edit the data loader in **data_loader.py**.  Instead, create a new file named **data_loader_val.py** containing the code for obtaining the data loader for the validation data.  You can access:
- the validation images at filepath `'/opt/cocoapi/images/train2014/'`, and
- the validation image caption annotation file at filepath `'/opt/cocoapi/annotations/captions_val2014.json'`.

The suggested approach to validating your model involves creating a json file such as [this one](https://github.com/cocodataset/cocoapi/blob/master/results/captions_val2014_fakecap_results.json) containing your model's predicted captions for the validation images.  Then, you can write your own script or use one that you [find online](https://github.com/tylin/coco-caption) to calculate the BLEU score of your model.  You can read more about the BLEU score, along with other evaluation metrics (such as TEOR and Cider) in section 4.1 of [this paper](https://arxiv.org/pdf/1411.4555.pdf).  For more information about how to use the annotation file, check out the [website](http://cocodataset.org/#download) for the COCO dataset.